### PR TITLE
add JitTransform

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Changes:
1. Add JitTransform: callback to run torch.compile/thunder.jit on model
2. Add test for JitTransform callback

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/jit_transform.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    Example:
+        >>> from nemo.lightning.pytorch.callbacks import JitTransform
+        >>> trainer = Trainer(callbacks=[JitTransform('torch)])
+    """
+
+    def __init__(self, backend: Optional[int] = None):
+        if backend is not None:
+            assert isinstance(backend, str)
+            backend = backend.lower()
+            assert backend in ['torch', 'thunder']
```
[the remark this answers is not in the document]
Comment (tfogal) [reply nemo/lightning/pytorch/callbacks/jit_transform.py:40]: This was probably for my benefit (Thunder dev here) to help understand how all the pieces fit together. It is certainly useful in that regard :-)

File: nemo/lightning/pytorch/callbacks/jit_transform.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+        if self.backend is None:
+            return
+        if self.backend == 'torch':
+            if hasattr(pl_module, 'module'):
+                trainer.lightning_module.module = torch.compile(pl_module.module)
+            elif hasattr(pl_module, 'model'):
+                trainer.lightning_module.model = torch.compile(pl_module.model)
+            else:
+                raise ValueError("Expected lightning_module to have a .model or .module attr.")
```
Comment (tfogal) [inline nemo/lightning/pytorch/callbacks/jit_transform.py:52]: as per @ananthsub 's comment, might be better to either add thunder into this PR from the get-go, or assert/raise NotImplemented here if `self.backend` is `thunder`
🤷 probably won't stay this way for long though so shouldn't matter much.

File: nemo/lightning/pytorch/callbacks/jit_transform.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akoumpa) [reply nemo/lightning/pytorch/callbacks/jit_transform.py:49]: Thank you guys, added the thunder branch.

File: nemo/lightning/pytorch/callbacks/__init__.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from nemo.lightning.pytorch.callbacks.ddp_parity_checker import DdpParityChecker
 from nemo.lightning.pytorch.callbacks.debugging import ParameterDebugger
 from nemo.lightning.pytorch.callbacks.garbage_collection import GarbageCollectionCallback
+from nemo.lightning.pytorch.callbacks.jit_transform import JitTransform
```
Comment (oyilmaz-nvidia) [inline nemo/lightning/pytorch/callbacks/__init__.py:18]: I am not sure if we add the model accelerators as callbacks into the Nemo. Let's chat.

**Review by oyilmaz-nvidia — CHANGES_REQUESTED** [review]
Let's chat about the design before merging the PR. Thanks!

File: nemo/lightning/pytorch/callbacks/jit_transform.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        if self.backend is None:
+            return
+        attr_name = extract_module_attr_name(pl_module)
+        model = getattr(pl_module, attr_name)
+        if self.backend == 'torch':
+            jit_model = torch.compile(model)
+        elif self.backend == 'thunder':
+            import thunder
+
+            jit_model = thunder.jit(model)
```
Comment (tfogal) [inline nemo/lightning/pytorch/callbacks/jit_transform.py:62]: I'd recommend using ThunderFX--this would mean
```python
import thunder
import thunder.dynamo
b_end = thunder.dynamo.ThunderCompiler()
jit_model = torch.compile(model, backend=b_end)
```

instead.  ThunderFX induces some overhead but we've found it to support a wider variety of models.

Comment (akoumpa) [reply nemo/lightning/pytorch/callbacks/jit_transform.py:62]: @tfogal feel free to refine this later, just want to get the bulletprint in.

File: nemo/lightning/pytorch/callbacks/jit_transform.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        else:
+            raise ValueError("got unexpected backend")
+        setattr(trainer.lightning_module, attr_name, jit_model)
```
Comment (ananthsub) [inline nemo/lightning/pytorch/callbacks/jit_transform.py:65]: does checkpointing "just work" if substituting in a jit-ed model like this?

Comment (akoumpa) [reply nemo/lightning/pytorch/callbacks/jit_transform.py]: switched to `.compile` from the nn.Moduel to avoid altering module's structure.

**Review by dimapihtar — DISMISSED** [review]
LGTM. Thank you!

**Comment by pablo-garay** [comment]
CI passed & only needed minimal conflict be solved
image
